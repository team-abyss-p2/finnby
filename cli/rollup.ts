import { dirname, join, relative } from "path";

import type { InputOptions, OutputOptions } from "rollup";
import replace from "@rollup/plugin-replace";
import { babel } from "@rollup/plugin-babel";
import { nodeResolve } from "@rollup/plugin-node-resolve";
import commonjs from "@rollup/plugin-commonjs";

import postcss from "postcss";
import modules from "postcss-modules";

import { renderChunk } from "./ssr";
import { Config } from "./config";
import { styledPlugin } from "./babel";
import { defaultPostCssPlugins } from "./postcss";

export type RollupConfig = InputOptions & {
    output: OutputOptions;
};

export function makeRollupConfig(config: Config, path: string): RollupConfig {
    const extractedStyles: string[] = [];
    const stylesheets = new Set<string>();
    const uid = `_${path.replace(/[^a-z0-9]+/g, "_")}`;

    return config.rollup({
        input: join(config.componentsDir, path),

        plugins: [
            replace({
                "process.env.NODE_ENV": JSON.stringify(
                    process.env.NODE_ENV || "development",
                ),
                preventAssignment: true,
            }),
            babel({
                babelHelpers: "bundled",
                extensions: [".js", ".jsx", ".ts", ".tsx"],
                exclude: /node_modules/,

                plugins: [
                    "@babel/plugin-transform-class-properties",
                    "@babel/plugin-transform-nullish-coalescing-operator",
                    styledPlugin({ uid, styles: extractedStyles }),
                ],
                presets: [
                    "@babel/preset-typescript",
                    [
                        "@babel/preset-react",
                        {
                            // Force the classic runtime since react
                            // cannot be auto-required from JS files
                            runtime: "classic",
                        },
                    ],
                    [
                        "@babel/preset-env",
                        {
                            modules: false,
                            targets: {
                                node: "17.0.0", // P2CE V8 Version: 9.5.172.19
                            },
                        },
                    ],
                ],
            }),
            nodeResolve({
                resolveOnly: ["classnames"],
                browser: false,
            }),
            commonjs(),
            {
                name: "finnby-postcss",
                async transform(code, id) {
                    if (!id.endsWith(".css")) {
                        return null;
                    }

                    let result: any;

                    const processor = postcss(
                        config.postcss([
                            modules({
                                getJSON: (_: any, json: any) => {
                                    result = json;
                                },
                            }),
                            ...defaultPostCssPlugins(),
                        ]),
                    );

                    const { css } = await processor.process(code, { from: id });

                    const fileName = relative(config.componentsDir, id);
                    stylesheets.add(fileName);

                    this.emitFile({
                        type: "asset",
                        name: `styles/${fileName}`,
                        source: css,
                    });

                    return {
                        code: `export default ${JSON.stringify(result)};`,
                    };
                },
            },
            {
                name: "finnby-output",
                async generateBundle(options, bundle) {
                    for (const file of Object.values(bundle)) {
                        if (file.type === "chunk" && file.isEntry) {
                            if (extractedStyles.length > 0) {
                                const processor = postcss(
                                    config.postcss(defaultPostCssPlugins()),
                                );

                                const code = extractedStyles.join("\n");
                                const { css } = await processor.process(code, {
                                    from: file.fileName,
                                });

                                this.emitFile({
                                    type: "asset",
                                    name: join(
                                        "styles",
                                        dirname(path),
                                        `finnby/${file.name}.css`,
                                    ),
                                    source: css,
                                });

                                stylesheets.add(
                                    join(
                                        dirname(path),
                                        `finnby/${file.name}.css`,
                                    ),
                                );
                            }

                            this.emitFile({
                                type: "asset",
                                name: join(
                                    "layout",
                                    dirname(path),
                                    `${file.name}.xml`,
                                ),
                                source: renderChunk(file, stylesheets),
                            });
                        }
                    }
                },
            },
        ],

        external: ["react", "@team-abyss-p2/finnby"],

        output: {
            dir: config.outDir,
            assetFileNames: "[name][extname]",
            entryFileNames: join(
                "scripts/components",
                dirname(path),
                "[name].js",
            ),
            format: "iife",

            name: uid,
            footer: `UiToolkitAPI.GetGlobalObject().FinnbyRuntime._mount(${uid}, $.GetContextPanel())`,

            globals: {
                react: "UiToolkitAPI.GetGlobalObject().React",
                "@team-abyss-p2/finnby":
                    "UiToolkitAPI.GetGlobalObject().FinnbyRuntime",
            },
        },
    });
}
