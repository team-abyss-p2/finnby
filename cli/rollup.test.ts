import { normalize } from "path";
import { OutputAsset, OutputChunk, rollup } from "rollup";

import { makeRollupConfig } from "./rollup";

interface RollupConfig<T> {
    input: string;
    hasRollup: boolean;
    hasPostCSS: boolean;
    chunks: T;
}

type RollupResult<T> = {
    [K in keyof T]?: OutputChunk | OutputAsset;
};

async function runRollup<T>(test: RollupConfig<T>): Promise<RollupResult<T>> {
    const componentsDir = "cli/__fixtures__/components";
    const outDir = "cli/__fixtures__/output";
    const postcss = jest.fn((config) => config);
    const rollupWrapper = jest.fn((config) => config);
    const path = test.input;

    const config = makeRollupConfig(
        {
            componentsDir,
            outDir,
            postcss,
            rollup: rollupWrapper,
        },
        path,
    );

    const build = await rollup(config);

    if (test.hasRollup) {
        expect(rollupWrapper).toHaveBeenCalled();
    } else {
        expect(rollupWrapper).not.toHaveBeenCalled();
    }

    if (test.hasPostCSS) {
        expect(postcss).toHaveBeenCalled();
    } else {
        expect(postcss).not.toHaveBeenCalled();
    }

    const { output } = await build.generate(config.output);

    const result: any = {};

    for (const [key, path] of Object.entries(test.chunks)) {
        result[key] = output.find(
            (chunk) => chunk.name && normalize(chunk.name) === normalize(path),
        );
    }

    return result;
}

describe("rollup", () => {
    /* eslint-disable @typescript-eslint/no-non-null-assertion*/

    it("should build simple components", async () => {
        const chunks = await runRollup({
            input: "component.tsx",
            hasRollup: true,
            hasPostCSS: false,
            chunks: {
                component: "component",
                layout: "layout/component.xml",
                stylesheet: "styles/finnby/component.css",
            },
        });

        expect(chunks.component).not.toBeUndefined();
        expect(chunks.layout).not.toBeUndefined();
        expect(chunks.stylesheet).toBeUndefined();

        expect(chunks.component!.type).toBe("chunk");
        expect(chunks.layout!.type).toBe("asset");

        expect((chunks.component as OutputChunk).code).toMatchSnapshot();
        expect((chunks.layout as OutputAsset).source).toMatchSnapshot();
    });

    it("should process stylesheet imports", async () => {
        const chunks = await runRollup({
            input: "stylesheet.tsx",
            hasRollup: true,
            hasPostCSS: true,
            chunks: {
                component: "stylesheet",
                layout: "layout/stylesheet.xml",
                stylesheet: "styles/finnby/stylesheet.css",
                userStyles: "styles/stylesheet.css",
            },
        });

        expect(chunks.component).not.toBeUndefined();
        expect(chunks.layout).not.toBeUndefined();
        expect(chunks.stylesheet).toBeUndefined();
        expect(chunks.userStyles).not.toBeUndefined();

        expect(chunks.component!.type).toBe("chunk");
        expect(chunks.layout!.type).toBe("asset");
        expect(chunks.userStyles!.type).toBe("asset");

        expect((chunks.component as OutputChunk).code).toMatchSnapshot();
        expect((chunks.layout as OutputAsset).source).toMatchSnapshot();
        expect((chunks.userStyles as OutputAsset).source).toMatchSnapshot();
    });

    it("should extract classes from styled components", async () => {
        const chunks = await runRollup({
            input: "styled.tsx",
            hasRollup: true,
            hasPostCSS: false,
            chunks: {
                component: "styled",
                layout: "layout/styled.xml",
                stylesheet: "styles/finnby/styled.css",
            },
        });

        expect(chunks.component).not.toBeUndefined();
        expect(chunks.layout).not.toBeUndefined();
        expect(chunks.stylesheet).not.toBeUndefined();

        expect(chunks.component!.type).toBe("chunk");
        expect(chunks.layout!.type).toBe("asset");
        expect(chunks.stylesheet!.type).toBe("asset");

        expect((chunks.component as OutputChunk).code).toMatchSnapshot();
        expect((chunks.layout as OutputAsset).source).toMatchSnapshot();
        expect((chunks.stylesheet as OutputAsset).source).toMatchSnapshot();
    });
});
