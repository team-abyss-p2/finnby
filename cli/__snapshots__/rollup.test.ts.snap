// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rollup should build simple components 1`] = `
"var _component_tsx = (function () {
    'use strict';

    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    function UserComponent() {
      return null;
    }

    return UserComponent;

}());
UiToolkitAPI.GetGlobalObject().FinnbyRuntime._mount(_component_tsx, $.GetContextPanel())
"
`;

exports[`rollup should build simple components 2`] = `"<root><scripts><include src=\\"file://{resources}/scripts/components/component.js\\"></include></scripts><Panel></Panel></root>"`;

exports[`rollup should extract classes from styled components 1`] = `
"var _styled_tsx = (function (React, finnby) {
    'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

    const StyledPanel = /*#__PURE__*/finnby.styled.static.Panel(\\"_styled_tsx_0_StyledPanel\\");
    const StyledButton = /*#__PURE__*/finnby.styled.static.Button(\\"_styled_tsx_1_StyledButton\\");
    const StyledLabel = /*#__PURE__*/finnby.styled.static(finnby.Label, {})(\\"_styled_tsx_3_StyledLabel\\"); // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types

    function UserComponent() {
      return /*#__PURE__*/React__default['default'].createElement(StyledPanel, {
        class: \\"otherClass\\"
      }, /*#__PURE__*/React__default['default'].createElement(StyledButton, null, /*#__PURE__*/React__default['default'].createElement(StyledLabel, {
        style: {
          paddingTop: \\"12px\\"
        }
      })));
    }

    return UserComponent;

}(UiToolkitAPI.GetGlobalObject().React, UiToolkitAPI.GetGlobalObject().FinnbyRuntime));
UiToolkitAPI.GetGlobalObject().FinnbyRuntime._mount(_styled_tsx, $.GetContextPanel())
"
`;

exports[`rollup should extract classes from styled components 2`] = `"<root><styles><include src=\\"file://{resources}/styles/finnby/styled.css\\"></include></styles><scripts><include src=\\"file://{resources}/scripts/components/styled.js\\"></include></scripts><Panel><Panel class=\\"otherClass _styled_tsx_0_StyledPanel\\"><Button class=\\"_styled_tsx_1_StyledButton\\"><Label style=\\"padding-top:12px;\\" class=\\"_styled_tsx_3_StyledLabel\\"></Label></Button></Panel></Panel></root>"`;

exports[`rollup should extract classes from styled components 3`] = `
"._styled_tsx_0_StyledPanel{
    background-color: red;
}
._styled_tsx_1_StyledButton{background-color: white;}
@keyframes _styled_tsx_2_fadeIn{
    from {
        opacity: 0;
    }
    to {
        opacity: 1;
    }
}
._styled_tsx_3_StyledLabel{
    color: blue;
    animation: _styled_tsx_2_fadeIn 500ms;
}"
`;

exports[`rollup should process stylesheet imports 1`] = `
"var _stylesheet_tsx = (function (React, finnby) {
    'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

    var styles = {\\"userClass\\":\\"_userClass_1j71q_1\\"};

    function UserComponent() {
      return /*#__PURE__*/React__default['default'].createElement(finnby.Panel, {
        class: styles.userClass
      });
    }

    return UserComponent;

}(UiToolkitAPI.GetGlobalObject().React, UiToolkitAPI.GetGlobalObject().FinnbyRuntime));
UiToolkitAPI.GetGlobalObject().FinnbyRuntime._mount(_stylesheet_tsx, $.GetContextPanel())
"
`;

exports[`rollup should process stylesheet imports 2`] = `"<root><styles><include src=\\"file://{resources}/styles/stylesheet.css\\"></include></styles><scripts><include src=\\"file://{resources}/scripts/components/stylesheet.js\\"></include></scripts><Panel><Panel class=\\"_userClass_1j71q_1\\"></Panel></Panel></root>"`;

exports[`rollup should process stylesheet imports 3`] = `
"._userClass_1j71q_1 {
    color: red;
}
"
`;
